enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,592 @@
+- API stability
+- Additional bare metal specific functionality in future
+- Upstream (Metal3) implications
+
+## Design Details
+
+`cluster-baremetal-operator` is a new Second Level Operator (SLO)
+whose operand is a controller for the BareMetalHost resource and
+associated components from the Metal3 project. The below sections
+covers different areas of the design of this new SLO.
+
+### Standard SLO Behaviors
+
+As an SLO, the BMO is expected to adhere to the standard expected
+behaviours of SLO, including:
+
+1. The CBO image should be tagged with
+   `io.openshift.release.operator=true` and contain a `/manifests`
+   directory with all of the manifests it requires the CVO to apply,
+   along with an `image-references` file listing the images referenced
+   by those manifests that need to be included in the OpenShift
+   release image.
+2. Implement a `ClusterOperator` resource, including updating its
+   status field with the `Available/Progressing/Degraded` conditions,
+   operator version number, and any `relatedObjects`.
+
+While not required initially, other common SLO patterns can be
+considered in future:
+
+1. Implement an operator configuration resource, including
+   `OperatorSpec` and `OperatorStatus` (as per in openshift/api#125)
+2. Implement cluster-level configuration under `config.openshift.io`
+   (as per openshift/api#124)
+3. Expose a `/metrics` endpoint for Prometheus to be configured to
+   scrape (via a `ServiceMontitor`) and define any relevant Prometheus
+   alert rules based on those metrics.
+
+### "Not In Use" SLO Behaviors
+
+Unlike most other SLOs, CBO is not applicable to all cluster
+configurations. On clusters running on an infrastructure platform
+other than `baremetal` it should adhere to the emerging expected
+behaviors for "not in use" SLOs, including:
+
+1. Setting its `ClusterOperator` condition to `Disabled=true`,
+  `Available=true`, `Progressing=False` with appropriate messages.
+2. User interfaces should convey this disabled state differently than
+   a failure mode (e.g. by graying it out).
+
+Currently, insights-operator is the only other example of an SLO
+following this pattern of using a `Disabled` status. Other somewhat
+similar cases following different patterns include:
+
+- Image registry and samples have a `Removed` management state where
+  `Degraded=False`, `Progressing=False`, `Available=True` with
+  `Reason=currentlyUnmanaged`.
+- The cluster credentials operator has a `disabled` config map setting
+  that can be used to disable the operator and it then sets
+  ClusterOperator status conditions to `Degraded=False`,
+  `Progressing=False`, `Available=True` with
+  `Reason=OperatorDisabledByAdmin` for all three conditions.
+
+### cluster-baremetal-operator Details
+
+The CBO will:
+
+- Be a new `openshift/cluster-baremetal-operator` project.
+- Publish an image called `cluster-baremetal-operator`.
+- Use CVO run-level 30, so its manifests will be applied in parallel
+  with the MAO.
```

Comment:
Why do we want to run it in parallel instead of having the MAO launch it? We could move the cluster-api-provider-baremetal deployment into the new operator along with everything else, and then the MAO could go back to simply managing 1 Deployment with platform-specific content.

Reply:
I think we have settled on the approach were this SLO is launched by the CVO. The operator then puts itself in a Disabled state immediately when not running on a Baremetal Platform.